### PR TITLE
Assign columns position while creating table metadata information

### PR DESCRIPTION
**Pull request by seut**
Moves the (BWC) logic of assigning correct column position into the DocTableMetada instead of doing this only on information_schema iterables.
This ensures that a `Reference.position()` is already set with a valid value, thus making it non-nullable.
Additionally it fixes the correct order of object inner type position assignment on table creation.

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/general/information-schema.rst [hunk 1]
```diff
@@ -294,11 +294,11 @@ table::
...
     | locations         | information                    |  11 | object_array             |
-    | locations         | information['evolution_level'] |  12 | smallint                 |
-    | locations         | information['population']      |  13 | bigint                   |
+    | locations         | information['evolution_level'] |  13 | smallint                 |
+    | locations         | information['population']      |  12 | bigint                   |
     | locations         | inhabitants                    |   7 | object                   |
-    | locations         | inhabitants['description']     |   8 | text                     |
-    | locations         | inhabitants['interests']       |   9 | text_array               |
+    | locations         | inhabitants['description']     |   9 | text                     |
+    | locations         | inhabitants['interests']       |   8 | text_array               |
```
Comment (seut) [inline docs/general/information-schema.rst:301]: The positions for inner object types were not in the order of how the CREATE TABLE statement declared them, see https://github.com/crate/crate/blob/master/blackbox/test_docs.py#L182.

**Review by mfussenegger — COMMENTED** [review]
Glanced over it and looks good so far :+1:

File: server/src/main/java/io/crate/metadata/Reference.java [hunk 2]
```diff
@@ -60,7 +60,12 @@
 
     public Reference(StreamInput in) throws IOException {
         ident = new ReferenceIdent(in);
-        position = in.readOptionalVInt();
+        if (in.getVersion().before(Version.V_4_5_0)) {
```
[the remark this answers is not in the document]
Comment (seut) [reply server/src/main/java/io/crate/metadata/Reference.java:63]: ah yes forgot to add a comment here. planned to adjust this before merge after the version is bumped

**Review by mfussenegger — APPROVED** [review]
nice :+1:

**Pull request merged**